pkg/registry/core/service/storage/rest.go

```diff
@@ -199,14 +199,25 @@ func (rs *REST) Create(ctx context.Context, obj runtime.Object, createValidation
 	nodePortOp := portallocator.StartOperation(rs.serviceNodePorts, dryrun.IsDryRun(options.DryRun))
 	defer nodePortOp.Finish()
 
+	needHealthCheck := apiservice.NeedsHealthCheck(service)
+	healthPort := int(service.Spec.HealthCheckNodePort)
 	if service.Spec.Type == api.ServiceTypeNodePort || service.Spec.Type == api.ServiceTypeLoadBalancer {
+		// If user has specified HealthCheck NodePort, we need to check if it conflicts with
+		// allocated ports.
+		// If there is no conflicts, then we reserve the given HealthCheck NodePort
+		// to prevent port allocator allocating the same port for Service NodePort later.
+		if needHealthCheck && healthPort > 0 {
+			if err := allocateHealthCheckNodePort(service, nodePortOp); err != nil {
```

Comment:
doesn't this create the same issue in reverse, where an auto-allocated health port could try to grab an explicit user-specified node port?

also, putting this inside a `if service.Spec.Type == api.ServiceTypeNodePort || service.Spec.Type == api.ServiceTypeLoadBalancer` block leaves a logical hole where the health port is not allocated if all of the following are true:

* needsHealthCheck == true
* healthPort > 0
* service.Spec.Type != api.ServiceTypeNodePort
* service.Spec.Type != api.ServiceTypeLoadBalancer

Reply:
> doesn't this create the same issue in reverse

when allocating health port automatically, apiserver would call `nodePortOp.AllocateNext()` which skips the allocated ports https://github.com/kubernetes/kubernetes/blob/c780554a64fcab1c23d0e2d44d601e02db9a4036/pkg/registry/core/service/storage/rest.go#L644-L649

> leaves a logical hole where the health port is not allocated if all of the following are true:

HealthCheckNodePort can only be set on LoadBalancer service with ExternalTrafficPolicy=Local

Reply:
> HealthCheckNodePort can only be set on LoadBalancer service with ExternalTrafficPolicy=Local

it's not obvious from this code that all cases are covered... should we move the health port allocation above the `if service.Spec.Type == api.ServiceTypeNodePort || service.Spec.Type == api.ServiceTypeLoadBalancer` block?

Reply:
hmmm 
1. healthCheckNodePort should make no effect if the type of Service is not `LoadBalancer` and `ExternalTrafficPolicy` is not `Local`
2. the validation will fail if users specify healthCheckNodePort on a Service other than `LoadBalancer` Service
3.  `apiservice.NeedsHealthCheck` always returns false if the type of Service is neither `LoadBalancer` nor `NodePort` https://github.com/kubernetes/kubernetes/blob/c780554a64fcab1c23d0e2d44d601e02db9a4036/pkg/api/service/util.go#L70-L86

Reply:
previously, we unconditionally did this:
```
	// Handle ExternalTraffic related fields during service creation.
	if apiservice.NeedsHealthCheck(service) {
		if err := allocateHealthCheckNodePort(service, nodePortOp); err != nil {
			return nil, errors.NewInternalError(err)
		}
	}
```

it seems good to keep the new code equally unconditional, e.g.:

```go
	// reserve the explicitly specified health check port before auto-allocating any node ports
	if apiservice.NeedsHealthCheck(service) && service.Spec.HealthCheckNodePort != 0 {
		if err := allocateHealthCheckNodePort(service, nodePortOp); err != nil {
			return nil, errors.NewInternalError(err)
		}
	}

	if service.Spec.Type == api.ServiceTypeNodePort || service.Spec.Type == api.ServiceTypeLoadBalancer {
		if err := initNodePorts(service, nodePortOp); err != nil {
			return nil, err
		}
	}

	// Handle ExternalTraffic related fields during service creation.
	// auto-allocate the health check port
	if apiservice.NeedsHealthCheck(service) && service.Spec.HealthCheckNodePort == 0 {
		if err := allocateHealthCheckNodePort(service, nodePortOp); err != nil {
			return nil, errors.NewInternalError(err)
		}
	}
```

Reply:
also, do we have a similar issue inside initNodePorts with auto-allocated node ports for items early in the `ports` list taking explicitly specified ports later in the list in a case like this:

```
spec:
  ports:
  - port: 1111
  - port: 2222
    nodePort: 30000
```

it really seems like we need a two-pass approach that allocates explicitly specified ports, then auto-allocates remaining ports

Reply:
It is already midnight here, I would fix that tomorrow :)

Reply:
The nodePort allocation logic is a bit complicated, I may need some time to sort it out :/